Security review PR - Collection of all "team project" feature related changes

pieterlukasse opened the pull request:
Jira Ticket: [VADC-951](https://ctds-planx.atlassian.net/browse/VADC-951)

### New Features
- "team project" based authorization, integrating with Gen3 Arborist. See also https://github.com/vinci-ohdsi/AtlasRoleConf/blob/601bd3c361dd38186d8246d8bd870146788480c2/README.md

### Breaking Changes
- none. The feature is activated through a new (optional) configuration option.

### Bug Fixes
- fixes an issue where sessions where unnecessarily stopped (https://github.com/uc-cdis/WebAPI/pull/107/commits/32e90bc0df31ab692c3167ec9c17ba500780821a)

### Improvements
- improves logging in various places by adding many .debug() calls (for DEBUG log level)

### Dependency updates
- the new feature depends on Arborist

### Deployment changes
- to enable this feature, set `security.ohdsi.custom.authorization.mode="teamproject"`. 

[VADC-951]: https://ctds-planx.atlassian.net/browse/VADC-951?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

pieterlukasse committed:
feat: prototype for extracting teamProject from initial request

...and setting that as a user role.

This prototype is still missing
a method to validate the teamProject
against Arborist.

feat: introduce custom configuration option

pieterlukasse committed:
feat: basic working version

pieterlukasse committed:
feat: extra checks on "service" and "method" parts of the authorization

pieterlukasse committed:
Update pom.xml with a better default authorization url

pieterlukasse committed:
temp test

pieterlukasse committed:
Revert "temp test"

This reverts commit 10488c629c08dd77466c7d7815beaf2adca452a1.

pieterlukasse committed:
feat: improve logging of jwt

pieterlukasse committed:
wip: working version tying permissions to teamproject role

pieterlukasse committed:
wip: add "Atlas users" role as default role when logged-in teamproject mode

pieterlukasse committed:
fix: add "Atlas users" as default system role

pieterlukasse committed:
wip: add more log statements for PermissionManager

pieterlukasse committed:
tmp: add default roles on the fly

... just for testing at the moment. System roles should really be
assigned to the user beforehand as part of the onboarding process...

pieterlukasse committed:
feat: ensure /user/me endpoint also triggers the UPDATE_TOKEN filter

pieterlukasse committed:
feat: ensure the teamproject is stored per user

...and allow reading current teamproject from cache in case of a
request to /user/refresh endpoint

pieterlukasse committed:
wip: moved logic to new filter class TeamProjectBasedAuthorizingFilter

pieterlukasse committed:
wip: remove commented-out code

pieterlukasse committed:
fix: working version with roles 15 and 1000 added

... these roles turned out to be essential for getting the error resolved (an error stating
"you are not authorized to view the concept set expression")

pieterlukasse committed:
wip: try adding only system role 15

pieterlukasse committed:
wip: improve logging

pieterlukasse committed:
wip: add log statement to Source classes

pieterlukasse committed:
wip: improve log statements

pieterlukasse committed:
fix: ensure reset of roles always happens

pieterlukasse committed:
fix: further fix of the roles reset logic

pieterlukasse committed:
fix: avoid double deletes

pieterlukasse committed:
wip: testing amazoncorreto alternative to deprecated openJDK 8

AidanHilt committed:
Let's try this

pieterlukasse committed:
wip: trying update-ca-trust

pieterlukasse committed:
fix: adding both session and user login as keys to the teamProject map

...to avoid the scenario where the user comes back in another session and still gets the same
teamProject role assigned...while in reality he might have been removed
from that team for example

pieterlukasse committed:
fix: making two internal methods private methods

pieterlukasse committed:
fix: remove session.stop() call from UpdateAccessTokenFilter

...and therefore from the flow of endpoints like /user/refresh.
Not sure why this was added there, as the /user/logout should be
the place to remove a session.
This solves a org.apache.shiro.subject.support.DisabledSessionException.

If the worry is that logout won`t be called, then the
expiry time should just be set to a short period.

The adjustment in JwtAuthRealm.java was to deal with the side
effect that occurred after the removal of the .stop in
the UpdateAccessTokenFilter filter:
java.lang.ClassCastException: io.buji.pac4j.subject.Pac4jPrincipal cannot be cast to java.lang.String

pieterlukasse committed:
fix: do not create a new session when requesting current session

pieterlukasse committed:
fix: synchronize part that adds are removes roles on the fly

...to avoid any race conditions where two threads could end up
trying to remove the same role for example.

pieterlukasse commented on src/main/java/org/ohdsi/webapi/security/PermissionService.java:
```diff
@@ -285,9 +285,9 @@ public boolean hasReadAccess(CommonEntity entity) {
-		} else {
+                // if (Objects.equals(authorizationInfo.getUserId(), entity.getCreatedBy().getId())){
+                //     hasAccess = true; // the role is the one that created the artifact - Is this for backwards compatibiliy mode? Should we write a migration instead?
+		        // } else {
```
todo: add a line similar to this https://github.com/uc-cdis/WebAPI/pull/107/files#diff-90975f196358543a36095f13ea6d0cfaaaeb80ac8b9a3a2adfcf896ec5baa628R61 for backwards compatibility.

pieterlukasse replied:
done

k-burt-uch commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,231 @@
+      Set<String> newDefaultRoles = new HashSet<String>(self.defaultRoles);
+
+      // if found, add teamproject as a role in the newUserRoles list:
+      if (teamProjectRole != null) {
```
```suggestion
      if (teamProjectRole != null || !teamProjectRole.isEmpty()) {
```

`    String currentTeamProjectName = this.authorizer.getCurrentTeamProjectRoleForCurrentUser() != null ? this.authorizer.getCurrentTeamProjectRoleForCurrentUser().getName() : "";
`
From line 174 - its possible for teamProjectRole to wind up as an empty string, which will pass the null check on 105. Since empty string won't be in the keyset, this should be fine, but it'd be safer long term to return false on empty string like you do null.

pieterlukasse committed:
fix: check teamProjectRole on null and empty string

Co-authored-by: burtonk <117617405+k-burt-uch@users.noreply.github.com>

pieterlukasse committed:
fix: add extra validation for team name

...and extra failsafe to avoid empty string names in the request parameters

pieterlukasse committed:
fix: improve validation and error message

pieterlukasse committed:
fix: leave the existing logic of hasReadAccess intact in PermissionService

...with the access rules only changing if this.authorizationMode.equals("teamproject").
For teamproject mode we want the users to only see items that are granted through a role
and not include the ones they have created (as they might have created this while
logged/in / working on another teamproject

pieterlukasse committed:
feat: remove unnecessary comments in Dockerfile

pieterlukasse commented on src/main/java/org/ohdsi/webapi/shiro/PermissionManager.java:
```diff
@@ -141,14 +179,34 @@ public void clearAuthorizationInfoCache() {
+  }
+
+  @Transactional
+  private void registerUser(String login, String name, Set<String> defaultRoles, Set<String> newUserRoles,
+      boolean resetRoles) {
+        registerUser(login, name, UserOrigin.SYSTEM, defaultRoles, newUserRoles, resetRoles);
+  }
```
not used - remove.

pieterlukasse replied:
done

pieterlukasse committed:
feat: remove unnecessary method from PermissionManager

pieterlukasse commented on src/main/java/org/ohdsi/webapi/shiro/PermissionManager.java:
```diff
@@ -176,18 +236,49 @@ public UserEntity registerUser(final String login, final String name, final User
+  private void addRolesForUser(String login, UserOrigin userOrigin, UserEntity user, Set<String> roles, boolean isSystemRole) {
+    if (roles != null) {
+      for (String roleName: roles) {
+        // Temporary patch/workaround (in reality the role should have been added by sysadmin?):
+        if (!isSystemRole) {
+          pocAddUserRole(roleName);
+        }
+        // end temporary patch
```
remove?

pieterlukasse replied:
as of now, this will only be called as part of `updateUser`, as part of `TeamProjectBasedAuthorizingFilter`, which only calls it after Arborist check on the (single) team project role... so it is not a concern in practice. But we need to decide whether we want roles being created on the fly in any way here...maybe not.

pieterlukasse commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,246 @@
+        }
+        // add teamproject role and related system role that
+        // enables read restrictions/permissions based read access configurations:
+        newDefaultRoles.add("read restricted Atlas Users"); // aka reserved system role 15 - TODO: since this is always added...maybe it should be one of the default roles instead...
```
to be discussed.

rkboyce replied:
I agree with you about changing this so that default roles are configurable and we don't need this here. THe default permissions seem to be set in src/main/java/org/ohdsi/webapi/shiro/management/AtlasSecurity.java and could be configured for both team projects and shared artifact owner/reader permissions

pieterlukasse commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,246 @@
+      }
+    }
+
+    // try to find "teamproject" param in url itself (there will be no redirectUrl if user session is still valid):
+    logger.debug("Fallback1: Looking for teamproject in request: {}....", url);
+    String[] teamProjectParams = getParameterValues(request, "teamproject");
+    if (teamProjectParams != null) {
+      logger.debug("Parameter teamproject found. Parsing....");
+      String teamProject = teamProjectParams[0].toLowerCase();
+      logger.debug("Found teamproject: {}", teamProject);
+      return teamProject;
+    }
+
+    logger.debug("Fallback2: Looking for teamproject in Action-Location header of request: {}....", url);
+    String actionLocationUrl = httpRequest.getHeader("Action-Location");
+    if (actionLocationUrl != null && actionLocationUrl.contains("teamproject=")) {
+      String teamProject = actionLocationUrl.split("teamproject=")[1];
+      logger.debug("Found teamproject: {}", teamProject);
+      return teamProject;
+    }
```
TODO - cleanup. These are likely not needed by now, and probably just dead code...

pieterlukasse replied:
or maybe not... some fallback scenarios might be needed in cases where the login was already made. The use cases logging in for first time vs opening again after changing team project selection probably need these different implementations.  Maybe not fallback 1 and 2...but probably at least one of them.

pieterlukasse reviewed (commented):
review done.

pieterlukasse committed:
feat: improve logging and comments on TeamProjectBasedAuthorizingFilter

pieterlukasse commented:
Some comments will receive follow-up in new tickets. 
The PR will be merged as part of https://github.com/uc-cdis/WebAPI/pull/101

pieterlukasse committed:
Merge branch 'pre_release_2.15_teamproject_dev' into feat/add_team_project_access_control

rkboyce commented on src/main/java/org/ohdsi/webapi/security/model/EntityPermissionSchema.java:
```diff
@@ -2,19 +2,27 @@
 import java.util.HashMap;
 import java.util.Map;
 
 public abstract class EntityPermissionSchema {
```
As we discussed, we are seeking to make the Team feature an extension to WebAPI that folks can opt for and configure. The core developers do not want the Team functionality written into the base WebAPI code (e.g., this class). We agree that the onInsert and onDelete methods will need a bit more logic than in you current method to allow for proper maintainance of shared artifact owner/reader permissions. Your current config-based approach decides on the concrete implementation in this abstract class for that functionality. An advantage of that is that it should apply generally to all the other artifacts through abstract class inheritance. It works for the Team permission management. That might be the right way to go for managing the shared artifact owner/reader permissions too. I write this because I think we can find a general way to assign those permissions that does not need to consider the specific artifact types. I looked at the other Java approaches - Factory, default methods, component instantiation, and optional methods. Factory would require allot of redesign and the other methods do not offer a nice general solution. So, I think this approach is best, possibly with some modifications to abstract out permission management.

pieterlukasse replied:
> I think this approach is best, possibly with some modifications to abstract out permission management.

Thanks. I'll keep it as is for now, and consider some further refinement in a follow-up PR, as I think I'll have to further change this part anyway soon.

rkboyce commented on src/main/java/org/ohdsi/webapi/security/model/EntityPermissionSchema.java:
```diff
@@ -2,19 +2,27 @@
     private final EntityType entityType;
     private final Map<String, String> readPermissions;
     private final Map<String, String> writePermissions;
 
```
There might actually be two different config options we are thinking of instead of one. The first is the external authorization, where a service checks if someone should have a role. The other would decide if the shared artifact owner/reader roles are used.

pieterlukasse replied:
Not sure if we need a config option for the latter. I think the mere presence of these roles, as part of the roles assigned to the user should trigger the logic.

rkboyce replied:
Ok - let's go with that and test it.

rkboyce commented on src/main/java/org/ohdsi/webapi/security/PermissionService.java:
```diff
@@ -64,6 +64,9 @@ public class PermissionService {
     private boolean securityEnabled;
 
+    @Value("${security.ohdsi.custom.authorization.mode}")
+    private String authorizationMode;
```
See my comment about possibly needing another config option for enabling shared artifact owner/reader permissions

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,253 @@
+          // in case of "teamproject" mode, we want all roles to be reset always, and
+          // set to only the one requested/found in the request parameters (following lines below):
+          String login = this.authorizer.getCurrentUser().getLogin();
+          boolean foundValidTeamProject = extractAndValidateTeamProjectRoleAndUpdateUserIfNecessary(this, login, request, response);
```
style: probably could remove the "IfNecessary" from the method name to shorten it (optional)

pieterlukasse replied:
I like the longer name, since otherwise it sounds like the method is always updating the user...I'll leave it for now and see if others complain. Perhaps the method should just be broken up into smaller ones that do just one thing... leaving for later, and I'll adjust if necessary.

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,253 @@
+
+    // synchronize on login to avoid race conditions (especially on DB updates) if
+    // this filter receives parallel requests for any reason:
+    synchronized (login.intern()) {
```
Question: does this work for both POST and GET? I think that the examples I have seen have the team project in the GET only but wonder if that could be an issue if folks can see and decide to copy URLS? For example, other people (hackers) seeing the url copy will learn about what teams exist.

pieterlukasse replied:
A malicious user can add any teamproject to the URL, there is no stopping that. We do however handle this case by validating whether the user has the right teamproject authorization (in Gen3 Arborist in this case).

rkboyce replied:
Ok, but do you all recommend the use POST or the header? if so, maybe we could get rid of the GET option? That might lower the chance of team project identifiers being copied on places where bots and hackers are sniffing.

pieterlukasse replied:
marking resolved, as discussed today.

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,253 @@
+      if (teamProjectRole != null && !teamProjectRole.trim().isEmpty()) {
+        // double check if this role has really been granted to the user:
+        if (self.checkGen3Authorization(teamProjectRole, login) == false) {
+          String errorMessage = "User is not authorized to access this team project's data";
```
Might consider a different HTTP response so as not to reveal that the Team exists e.g., 400 Bad Request

pieterlukasse replied:
Good point, although this 403 also returns for non-existing teams...so we've covered that.

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,253 @@
+        self.authorizer.updateUser(login, newDefaultRoles, newUserRoles, true);
+        return true;
+      } else {
+        String errorMessage = "The teamproject is compulsory when on authorizationMode==teamproject configuration";
```
same comment as above, maybe 400 Bad Request?

pieterlukasse replied:
see previous reply.

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,253 @@
+  private boolean checkGen3Authorization(String teamProjectRole, String login) throws Exception {
+    logger.debug("Checking Gen3 Authorization for 'team project'={} and user={} using service={}", teamProjectRole, login, this.authorizationUrl);
+    RestTemplate restTemplate = new RestTemplate();
+    String arboristAuthorizationURL = this.authorizationUrl;
```
The first paramater looks like it should be in the configuration file maybe

pieterlukasse replied:
which one? this.authorizationUrl? This does come from the pom or the properties file.

rkboyce replied:
Please ignore this comment. I think I was referring to the teamProjectRole but must have been confused since that is given to the WebAPI from Atlas.

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,253 @@
+    }
+  }
+
+
```
should there be a check for an encrypted channel to sending the authorization?

pieterlukasse replied:
good one. I agree this could be the default behavior (a valid https endpoint - with valid certificate). Not the case right now. 
TODO.

pieterlukasse replied:
Update: I think we should **_not_** enforce it (leaving it to each devops team to decide what's best), but if it is connecting to an external service via HTTPS, then it should fail if the certificate is invalid. 
I tested this and the current code already does the necessary validation (i.e. failing when certificate is self-signed for example). This is how I tested it:
- I've set the property `security.ohdsi.custom.authorization.url=https://reverse-proxy/auth/mapping` where `reverse-proxy` is running on a self-signed certificate. 
   - This resulted in a certificate related error ✅ :
 ```
java.lang.Exception: 
org.springframework.web.client.ResourceAccessException: 
I/O error on POST request for "https://reverse-proxy/auth/mapping": 
PKIX path building failed: 
sun.security.provider.certpath.SunCertPathBuilderException: 
unable to find valid certification path to requested target; 
nested exception is javax.net.ssl.SSLHandshakeException: 
PKIX path building failed: 
sun.security.provider.certpath.SunCertPathBuilderException: 
unable to find valid certification path to requested target
```

pieterlukasse committed:
fix: lowercase the login before passing it downstream

This copies the behavior also seen in UpdateAccessTokenFilter class

m0nhawk reviewed (approved):
👍

m0nhawk committed:
fix: convert email to lowercase

pieterlukasse committed:
Merge pull request #111 from uc-cdis/fix/lowercase-email

fix: convert email to lowercase

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,253 @@
+      }
+      JSONObject teamProjectAuthorization = teamProjectAuthorizations.getJSONObject(0);
+
+      // check if the authorization contains the right "service" and "method" values:
```
I agree - make this configurable

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,253 @@
+        return false;
+      }
+      JSONObject teamProjectAuthorization = teamProjectAuthorizations.getJSONObject(0);
+
```
to configuration?

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,253 @@
+  /**
+   * Tries to find the team project information in the given request or as part of the
+   * stored this.authorizer.getCurrentTeamProjectRoleForCurrentUser(). Returns null
+   * if nothing can be found.
```
Per my comment about the team project being exposed in the URL which might be copied and reveal Teams to hackers - should we put this in the request header or do a POST and just require that Atlas/WebAPI can only initiate this functionality from e.g., a coded (e.g., JS) http request?

pieterlukasse replied:
I'm not sure I understand the question.
But maybe my previous comments on this topic already clarify the code and the question does not apply anymore?

rkboyce replied:
I replied above. I am ok if you close the comment.

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,253 @@
+  private String extractTeamProjectFromRequestParameters(ServletRequest request) throws Exception {
+    // Get the url
+    HttpServletRequest httpRequest = (HttpServletRequest) request;
+    String url = httpRequest.getRequestURL().toString();
```
Is this where you allow for the authorization header to hold the team project string? That would probably address things.

pieterlukasse replied:
see previous comments. Please let me know if there is still an issue.

rkboyce replied:
I don't think it is an issue per your discussion.

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java:
```diff
@@ -0,0 +1,253 @@
+      }
+    }
+    logger.debug("Current teamproject: {}...", currentTeamProjectName);
+    logger.debug("Checking if a teamproject has been specified in the request...");
```
Is this case needed? If not, we should probably remove it for the security reasons mentioned above.

pieterlukasse replied:
see previous comments regarding the authorization check.

rkboyce replied:
I understand the authorization check but that would come after the GET had revealed the team. A user who successfully logs in might find the URL, copy it with the team, and paste it into a document that becomes public. This reveals the ID of the team project to others. That information reveal probably is not necessary and creates a small risk of hacking through methods such as perhaps social hacking.

pieterlukasse replied:
See also my previous reply on the concern regarding exposing the team project code. Furthermore, we are actually showing the team project code on the UI as well, so currently we are not trying to hide it at all.

pieterlukasse replied:
marking resolved, as discussed today.

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/filters/UpdateAccessTokenFilter.java:
```diff
@@ -115,11 +115,6 @@ protected boolean preHandle(ServletRequest request, ServletResponse response) th
 
     login = UserUtils.toLowerCase(login);
 
```
This looks like removal of unecessary code?

rkboyce commented on src/main/java/org/ohdsi/webapi/shiro/PermissionManager.java:
```diff
@@ -96,8 +109,32 @@ public void removeUserFromRole(String roleName, String login, UserOrigin origin)
+      logger.debug("Removing user from role: {}, {}, {}", user.getLogin(), role.getName(), userRole.getOrigin());
       this.userRoleRepository.delete(userRole);
+    }
+  }
```
As we build out the shared artifact owner/reader permissions, we will need to ensure that they are system roles since this and related code removes all non-system roles

rkboyce commented on src/main/java/org/ohdsi/webapi/source/SourceService.java:
```diff
@@ -111,13 +115,15 @@ public void checkConnection(Source source) {
     public Source getPrioritySourceForDaimon(SourceDaimon.DaimonType daimonType) {
 
         List<Source> sourcesByDaimonPriority = sourceRepository.findAllSortedByDiamonPrioirty(daimonType);
+        logger.debug("Found {} sources for daimon type {}.", sourcesByDaimonPriority.size(), daimonType.name());
```
Is it true that all coding changes in this file are for logging?

rkboyce commented:
@pieterlukasse  - I am done with my code review. Please note that I am not a security expert and I recommend that you have a security expert/group review the code after revision and perform the recommended security testing.

pieterlukasse committed:
fix: revert toLowerCase and use lower() in SQL query itself

pieterlukasse committed:
fix: take login from shiro-parsed principal instead of DB

... to avoid the issue where the login is all lowercase in db

pieterlukasse committed:
feat: move the defaultRoles definition into AtlasSecurity

pieterlukasse committed:
fix: move authorizationMode check to PostConstruct

...to avoid NullPointerException as attributes
are not yet wired when in constructor

pieterlukasse commented:
@rkboyce thanks for the review! I'm merging this PR into the pre_release/dev branch. New/extra changes will be made in follow-up PRs. I think the current PR is fine now, since remaining open comments are mostly about config options, default roles or reminder notes regarding future features like shared artifact owner/reader permissions. Let's revisit these in the next PR (if they are still applicable by then).
